src/hb-buffer.cc

```diff
@@ -1755,6 +1755,22 @@ hb_buffer_append (hb_buffer_t *buffer,
   memcpy (buffer->info + orig_len, source->info + start, (end - start) * sizeof (buffer->info[0]));
   if (buffer->have_positions)
     memcpy (buffer->pos + orig_len, source->pos + start, (end - start) * sizeof (buffer->pos[0]));
+
+  if (source->content_type == HB_BUFFER_CONTENT_TYPE_UNICODE)
+  {
+    if (start == 0 && orig_len == 0 && source->context_len[0] && !buffer->context_len[0])
+    {
+      buffer->context_len[0] = source->context_len[0];
+      for (unsigned i = 0; i < source->context_len[0]; i++)
+	buffer->context[0][i] = source->context[0][i];
+    }
+    if (end == source->len && source->context_len[1] && !buffer->context_len[1])
+    {
+      buffer->context_len[1] = source->context_len[1];
+      for (unsigned i = 0; i < source->context_len[1]; i++)
+	buffer->context[1][i] = source->context[1][i];
+    }
+  }
```

Comment:
This looked like the easiest way to make `verify_buffer_safe_to_break ()` copy the context when appending various buffer, but I’m not comfortable with the fact that it changes the behavior of public API.

Reply:
I'm fine with behavior change, as I filed https://github.com/harfbuzz/harfbuzz/issues/1843

Let me review correctness though.

Reply:
The logic seems like can use some fine-tuning to me.  See similar logic in add_utf. Ideally we should reuse that as I described in #1843. I'll see if I can help with that.

I'll comment above about what doesn't look right to me.

Reply:
That seems too much work. Let's duplicate logic.